File: src/etch/bindings/csharp/transport/AbstractStartable.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

namespace Etch.Transport
{
    /// <summary>
    /// Generic implementation of Startable
    /// </summary>
    abstract public class AbstractStartable : Startable
    {
        [MethodImpl(MethodImplOptions.Synchronized)]
        public void Start()
        {
            if (IsStarted())
                throw new Exception("Is already started");

            try
            {
                started = true;
                Start0();
            }
            catch (Exception e)
            {
                SetStopped();
                Stop0();
                throw e;
            }
        }

        [MethodImpl(MethodImplOptions.Synchronized)]
        public void Stop()
        {
            CheckIsStarted();

            started = false;

            Stop0();
        }

        private bool started;

        private void SetStarted()
        {
            started = true;
        }

        protected void SetStopped()
        {
            started = false;
        }

        public bool IsStarted()
        {
            return started;
        }

        /// <summary>
        /// Check that the startable is started
        /// Exception:
        ///     ThreadInterruptedException if not started
        /// </summary>
        /// 
        public void CheckIsStarted()
        {
            if (!IsStarted())
                throw new System.Threading.ThreadInterruptedException("is not started");
        }

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        abstract protected void Start0();

        /// <summary>
        /// 
        /// </summary>
        /// Exception:
        ///     throws Exception
        abstract protected void Stop0();

    }
}

File: src/etch/util/AbstractStartable.java
/* $Id$
 *
 * Created by wert on Mar 26, 2007.
 *
 * Copyright (c) 2007 Cisco Systems, Inc. All rights reserved.
 */

package etch.util;

/**
 * Generic implementation of Startable.
 */
abstract public class AbstractStartable implements Startable
{
	final public synchronized void start() throws Exception
	{
		if (isStarted())
			throw new IllegalStateException( "is already started" );
		
		try
		{
			started = true;
			start0();
		}
		catch ( Exception e )
		{
			setStopped();
			stop0();
			throw e;
		}
	}

	final public synchronized void stop() throws Exception
	{
		checkIsStarted();
		setStopped();
		stop0();
	}
	
	/**
	 * Signals (from a subclasser) that the startable has stopped.
	 */
	final protected void setStopped()
	{
		started = false;
	}
	
	final public boolean isStarted()
	{
		return started;
	}
	
	final public void checkIsStarted()
	{
		if (!isStarted())
			throw new IllegalStateException( "is not started" );
	}

	private boolean started;
	
	/**
	 * @throws Exception
	 */
	abstract protected void start0() throws Exception;
	
	/**
	 * @throws Exception
	 */
	abstract protected void stop0() throws Exception;
}
